http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -127,6 +127,62 @@ class SamAuthProvider[F[_]: OpenTelemetryMetrics](
     } yield (callerActions, projectCallerActions)
   }
 
+  /**
+   * Get all Sam resource IDs of the given type on which the user is granted supported permissions. Returns explicitly
+   * granted policies only; users with workspace owner access, for example, can see all runtimes in the workspace, but
+   * getAuthorizedIds[SamRuntimeResourceId] is not guaranteed to return all runtimes in that workspace. Works for any
+   * Sam resource type except SharedApp and similar alternative app types; these require a specific Sam resource object
+   * to determine permissions, as these are stored on the individual records.
+   * @param isOwner whether to return only resources that the user is granted ownership on (has `ownerRoleName`,
+   * and `ownerRoleName` is an element in `policyNames` on the SamResource)
+   * @param userInfo the current user
+   * @param samResource the resource archetype (App, Runtime, etc)
+   * @param decoder converts Json -> SamResourceId, needed by SamDao
+   * @param ev the request context
+   * @tparam R SamResourceId
+   * @return List[SamResourceId] of all resources of type R which are visible to the user at the given access level
+   */
+  override def getAuthorizedIds[R <: SamResourceId](
+    isOwner: Boolean,
+    userInfo: UserInfo
+  )(implicit
+    samResource: SamResource[R],
+    decoder: Decoder[R],
+    ev: Ask[F, TraceId]
+  ): F[List[R]] = {
+    val authHeader = Authorization(Credentials.Token(AuthScheme.Bearer, userInfo.accessToken.token))
+    val ownerRole: SamRole = samResource.ownerRoleName
+    val canOwnerRead: Boolean = samResource.policyNames.isEmpty || samResource.policyNames.exists {
+      policyName: SamPolicyName =>
+        ownerRole.asString == policyName.toString
+    }
+    for {
+      resourcesAndPolicies: List[(R, SamPolicyName)] <- samDao
+        .getResourcePolicies[R](authHeader, samResource.resourceType)
+      _ = logger.info(s"getAuthorizedIds : ${resourcesAndPolicies.mkString}")
+
+      authorizedPolicies =
+        if (isOwner && !canOwnerRead) {
```

Comment:
thanks for the helpful annotations!! Leo is slowly becoming more understandable, hopefully